### PR TITLE
8273300: Check Mutex ranking during a safepoint

### PR DESCRIPTION
**Pull request by coleenp**
This change checks lock ranking during a safepoint.  For some reason, safepoint checking was excluded, probably from the days where Safepoint_lock and Threads_lock were used.
Because of checking during a safepoint, some locks had to get lower ranks.  The CR has the details of which locks these were.  The Service_lock complicates things because it's held during oops_do, which may take out other G1 locks.
This was built and tested with Shenandoah.  Thanks to @zhengyu123 for the changes in Shenandoah.
Tests run tier1-8.

---------
### Progress
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue
- [x] Change must be properly reviewed

### Issue
 * [JDK-8273300](https://bugs.openjdk.java.net/browse/JDK-8273300): Check Mutex ranking during a safepoint

### Reviewers
 * [Erik Österlund](https://openjdk.java.net/census#eosterlund) (@fisk - **Reviewer**) ⚠️ Review applies to 133559726a69084c5e75991e6c573b9592656c57
 * [David Holmes](https://openjdk.java.net/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to 9920c227432067e03fcfc45db7f1d4ab6af71fa3
 * [Patricio Chilano Mateo](https://openjdk.java.net/census#pchilanomate) (@pchilano - Committer) ⚠️ Review applies to eec9b842e697965e53539ee5e85480614e30941d

### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.java.net/jdk pull/5467/head:pull/5467` \
`$ git checkout pull/5467`

Update a local copy of the PR: \
`$ git checkout pull/5467` \
`$ git pull https://git.openjdk.java.net/jdk pull/5467/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 5467`

View PR using the GUI difftool: \
`$ git pr show -t 5467`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.java.net/jdk/pull/5467.diff">https://git.openjdk.java.net/jdk/pull/5467.diff</a>

</details>

### CODE REVIEW

**Review by fisk — APPROVED** [review]
Looks good.

**Comment by coleenp** [comment]
Thanks Erik!

File: src/hotspot/share/memory/universe.cpp [hunk 1]
```diff
@@ -122,7 +122,6 @@ LatestMethodCache* Universe::_throw_illegal_access_error_cache = NULL;
 LatestMethodCache* Universe::_throw_no_such_method_error_cache = NULL;
 LatestMethodCache* Universe::_do_stack_walk_cache     = NULL;
 
-bool Universe::_verify_in_progress                    = false;
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/universe.cpp:125]: This cleanup seems completely unrelated to your mutex change and is best left to a separate cleanup RFE.

File: src/hotspot/share/memory/universe.cpp [hunk 2]
```diff
@@ -1113,7 +1107,6 @@ void Universe::verify(VerifyOption option, const char* prefix) {
     StringTable::verify();
   }
   if (should_verify_subset(Verify_CodeCache)) {
-    MutexLocker mu(CodeCache_lock, Mutex::_no_safepoint_check_flag);
```
Comment (dholmes-ora) [inline src/hotspot/share/memory/universe.cpp:1116]: Is this needed to allow the new rankings to work? And is this enabled by the _verify_in_progress change? If so I'd rather see all of that related stuff changed first in a separate RFE that can easily be independently backported.

**Review by dholmes-ora — COMMENTED** [review]
Hi Coleen,

The core change seems clean and simple, but the fanout of the ranking changes is far less clear and some of this looks like it could/should be separated out - see comments below.

If I see a lock with a rank service-1, then should I infer that lock can be acquired while the service (or notification) lock is held? And that if lock A is service-1 and lock B is service-2, then B can be acquired while holding A?

Thanks,
David

File: src/hotspot/share/memory/universe.cpp (hunk 2 above)
Comment (coleenp) [reply src/hotspot/share/memory/universe.cpp:1116]: Yes, this is needed.  This verification is done during a safepoint, so we don't need this lock.  The CodeCache_lock has a vary low ranking and takes out VtableStubs_lock which is a higher ranking.  With this change, we do not take out the CodeCache_lock, so it's needed for this change.  I see no reason whatsoever to backport it though.

File: src/hotspot/share/memory/universe.cpp (hunk 1 above)
Comment (coleenp) [reply src/hotspot/share/memory/universe.cpp]: Done.

**Comment by coleenp** [comment]
I just remerged from master and a couple of copyright updates leaked in from my script (which I shouldn't have used).

File: src/hotspot/cpu/zero/vm_version_zero.cpp [hunk 3]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline src/hotspot/cpu/zero/vm_version_zero.cpp:2]: Unwanted copyright date change.

File: src/hotspot/share/memory/universe.cpp (hunk 2 above)
Comment (dholmes-ora) [reply src/hotspot/share/memory/universe.cpp:1109]: This isn't connected to the _verify_in_progress change, so that is fine.

File: src/hotspot/share/runtime/mutex.hpp [hunk 4]
```diff
@@ -45,12 +45,12 @@ class Mutex : public CHeapObj<mtSynchronizer> {
...
   enum lock_types {
        event,
-       service        = event          +   3,
+       service        = event          +   6,
        stackwatermark = service        +   3,
        tty            = stackwatermark +   3,
        special        = tty            +   3,
        oopstorage     = special        +   3,
-       leaf           = oopstorage     +   2,
+       leaf           = oopstorage     +  10,
```
Comment (dholmes-ora) [inline src/hotspot/share/runtime/mutex.hpp:53]: Why do we need such a big gap here? Is there any reason we can't just use the same gap between all named rankings? As it is there seems to be no rationale for the  "+ N" value used.

File: src/hotspot/share/utilities/growableArray.hpp [hunk 5]
```diff
@@ -1,5 +1,5 @@
 /*
- * Copyright (c) 1997, 2020, Oracle and/or its affiliates. All rights reserved.
+ * Copyright (c) 1997, 2021, Oracle and/or its affiliates. All rights reserved.
```
Comment (dholmes-ora) [inline src/hotspot/share/utilities/growableArray.hpp:2]: Unwanted copyright change.

**Review by dholmes-ora — APPROVED** [review]
Hi Coleen,

It wasn't clear if you thought you had already fixed the incorrect copyright notices so I flagged the two I saw.

I still have some queries about this - see comment below - but nothing that would stop it from being pushed as-is.

Thanks,
David

File: src/hotspot/share/runtime/mutex.hpp (hunk 4 above)
Comment (coleenp) [reply src/hotspot/share/runtime/mutex.hpp:53]: The gap is pretty much arbitrary, but there's already some leaf-2 and having overlapping rankings is something we don't really want, so I changed it to 10 for now.

File: src/hotspot/share/utilities/growableArray.hpp (hunk 5 above)
Comment (coleenp) [reply src/hotspot/share/utilities/growableArray.hpp]: I think I finally reverted them.  My commit script fixed them back again.

File: src/hotspot/share/runtime/mutex.hpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/mutex.hpp:53]: Okay but is there a reason not to use the same gap?

Comment (coleenp) [reply src/hotspot/share/runtime/mutex.hpp:53]: If the gap is 2, then leaf-2 will == oopstorage?  I could make it 3 but it doesn't actually matter.

**Review by pchilano — APPROVED** [review]

File: src/hotspot/share/runtime/mutex.cpp [hunk 6]
```diff
@@ -372,11 +372,14 @@ Mutex* Mutex::get_least_ranked_lock_besides_this(Mutex* locks) {
 void Mutex::check_rank(Thread* thread) {
   Mutex* locks_owned = thread->owned_locks();
 
-  // We expect the locks already acquired to be in increasing rank order.
-  for (Mutex* tmp = locks_owned; tmp != NULL; tmp = tmp->next()) {
-    if (tmp->next() != NULL) {
-      assert(tmp->rank() < tmp->next()->rank()
+  if (!SafepointSynchronize::is_at_safepoint()) {
```
Comment (pchilano) [inline src/hotspot/share/runtime/mutex.cpp]: I think removing this conditional was okay in your previous change.

Comment (coleenp) [reply src/hotspot/share/runtime/mutex.cpp]: Thanks Patricio. I've removed it and retesting.

File: src/hotspot/share/runtime/mutex.hpp (hunk 4 above)
Comment (dholmes-ora) [reply src/hotspot/share/runtime/mutex.hpp:53]: Obviously the gap needs to be big enough to avoid the overlap, but the presence of +1, +3, +6 and +10 just raise questions as to why different values are used. Why not make them all +10 if you need 10 slots in some areas?

Comment (coleenp) [reply src/hotspot/share/runtime/mutex.hpp:53]: I'm going to address this with a future change because we don't have range checking for overlap yet, and they should all be some consistent and not arbitrary range.  If it was less arbitrary, I'd pick another 6 because
runtime/mutexLocker.cpp:  def(Metaspace_lock               , PaddedMutex  , leaf-3,      true,  _safepoint_check_never);
The rank 'service' had 6 because there are at least 4 locks below that.  Again, I should add range checking and maybe making them all 10 makes a lot of sense.  Locks don't nest that deeply.  I'll add this as a comment in: JDK-8176393.

**Comment by coleenp** [comment]
Thanks Patricio for the code review.  Thanks David for all the comments.
/integrate
